### DISCUSSION TITLE
2.2 Dev: Documentation Discussion

### DISCUSSION THREAD

**Discussion by tajmone** [AnssiR66/AlanStdLib#145] [StdLib Documentation]
This Issue is a thread for discussing and planning the documentation for the upcoming Library v2.2 release. These include tutorials, the _Alan Cookbook_ and any other documentation files.

For each tasks an issue will be created under the [2.2 Dev: Documentation] __milestone__ (or, in case of complex changes involving multiple steps, a dedicated __milestone__ might be created), and as part of the [StdLib 2.2 Dev] __project__. This should make it easier to track the progress by topics.

[2.2 Dev: Documentation]: https://github.com/AnssiR66/AlanStdLib/milestone/6
[StdLib 2.2 Dev]: https://github.com/AnssiR66/AlanStdLib/projects/2

**Comment by tajmone** [comment]
# Conversion of StdLib Man & Cookbook

I'll start to work locally on the conversion to AsciiDoc of the __Library Manual__ and the __Alan Cookbok__.

I'll just convert them as they are, without changing the contents — just some slight adapations to fit the AsciiDoc model.

Once they are ready to be used with the new automation system, you can then revise the code examples (which will be in Alan sources) so that everything mirrors the current state of the library.

From there onward, the toolchain will keep the documentation always aligned to the library state, and all transcripts will be auto-updated. If any changes in the library ever break the examples of the documentation, it will show up in Git as either unexpected changes or Alan compiler errors.

**Comment by AnssiR66** [comment]
OK, thanks! I will try to locate a docx version of the Cookbook somewhere for you.



________________________________

**Comment by tajmone** [comment]
# AsciiDoc to PDF: the XSL-FOP Template

I've proposed to @thoni56 that we move the [`xsl-fopub/`][xsl-fopub] folder to a repository of its own so it can be used by any project as Git submodule:

* https://github.com/alan-if/alan-docs/issues/44

We could then use the same Alan template of the [Alan-Docs] project to build the PDF version of the documentation (_StdLib Manual_, _Alan Cookbook_) by importing it as a Git submodule. This would mean that whenever the template is updated we can just move the submodule pointer and enjoy the improvements.

Overall, keeping the same documentation toolchain in Alan-related projects is going to be beneficial in the long term, since it will mean less maintainance work when third party assets are updated (Asciidoctor, FOP, etc.). With Alan being a small niche-community, sharing common assets is always preferable as it lessens the burden of keeping up with the evolution of dependcies.

Of course, the XSL-FOP template is crucial for the StdLib in order to continue to provide sldo a PDF version of the documentation (which probably many users expect to find). Adapting the XSL template to the needs of Alan documentation was quite hard work, for it's not an easy format to work with; so definitely we should reuse it.

From the StdLib perspective, the advantages of moving the documentation to AsciiDoc:

* Being text-based it's version controllable, therefore much easier to work on collaboratively.
* Being script driven, it allows dynamic contents generation (Alan sources and transcripts).
* The documentation could be made available online via GitHub pages. There are many tools to build websites from AsciiDoc.
* Asciidoctor supports lot's of output backends besided HTML and PDF, so in the future we (or anyone else) could also convert the documentation to other formats (ePub, Kindle, CHM, whatever).



[Alan-Docs]: https://github.com/alan-if/alan-docs
[xsl-fopub]: https://github.com/alan-if/alan-docs/tree/master/_assets/xsl-fopub

**Comment by AnssiR66** [comment]
Sounds good Tristano, we can do it the way you have suggested.

**Comment by tajmone** [comment]
# Reorganization of Extras Source/Dest Folders

Finally, today I've managed to find some time and peace of mind to prepare the ground for adding to the project the _StdLib Manual_ and the _ALAN Cookbook_ in AsciiDoc.

> __NOTE__ — sorry if I've done a few force-pushes on the dev branches, but after fast-forward merging (and pushing) I later decided to use a `--no-ff` merge instead, to keep a cleaner history graph of these branches and better understand how they are being integrated with each other.
> 
>  I hope you didn't pull in the meantime, and I apologyse for this.

I've edited the whole toolchain to support multiple documentation source/target folders, and reorganized the Extras subfolders as follows:

|      source folder      |    output folder    |      description       |
|-------------------------|---------------------|------------------------|
| `extras_src/tutorials/` | `extras/tutorials/` | _Clothing Guide_, etc. |

I can now add at any time subfolder for the _StdLib Man_ and _Cookbook_, and the toolchain car build them by simply tweaking a single script variable.

This also means that we'll be able to work on porting these two documents to AsciiDoc in dedicated branches, without editing the toolchain, and only merge them in the dev baseline when they are ready to go. I was thinking of some folder names along these lines:

|     source folder      |   output folder    |        description        |
|------------------------|--------------------|---------------------------|
| `extras_src/manual/`   | `extras/manual/`   | _Standard Library Manual_ |
| `extras_src/cookbook/` | `extras/cookbook/` | _ALAN Cookbook_           |

... or along those lines. If you have any preferences and suggestions for these folders name please advice me (changing them later on is not a problem though).

For more details, see:

- [Extras Src TODO](https://github.com/AnssiR66/AlanStdLib/blob/dev-2.2.0-docs/extras_src/TODO.md#group-docs-by-folders)
- [Extras Src README](https://github.com/AnssiR66/AlanStdLib/blob/dev-2.2.0-docs/extras_src/README.md#the-documentation-toolchain)

# StdLib Manual

As mentioned in a private email, I've been working locally on porting the Library Manual from PDF to AsciiDoc (quite lengthy work, have to restructure manually every paragraph). I'm not quite sure when I should start adding drafts of the ADoc port to the dev branches, i.e. if I should wait that I've converted the whole doc (even if in draft) or if I should just start commiting as I go along (so you can preview the ongoing work).

Any suggestions?
